Implement Parameter Bind on SQL Where clause

silathdiir opened the pull request:
Closes #275 

#### Description

Implements Parameter Bind for simple expression on SQL Where clause as below.

```
prepare fooplan (smallint) as select * from schema_name.table_name where col1 = $1
```

#### Is it a feature that change user experience?

Yes. Some database drivers could directly use Extended Protocol. And this PR updates the Bind logic for `Select` queries. 

#### Client output

```
psql (12.1, server 0.0.0)
Type "help" for help.

xxx=> create schema schema_name;
CREATE SCHEMA
xxx=> create table schema_name.table_name (col1 smallint, col2 smallint, col3 smallint);
CREATE TABLE
xxx=> insert into schema_name.table_name values (1, 2, 3);
INSERT 0 1
xxx=> prepare fooplan (smallint) as select * from schema_name.table_name where col1 = $1;
PREPARE
xxx=> execute fooplan(1);
 col1 | col2 | col3
------+------+------
    1 |    2 |    3
(1 row)
```

alex-dukhno commented on src/binder/src/lib.rs:
```diff
@@ -63,6 +56,34 @@ fn bind_insert(stmt: &mut Statement, params: &[Expr]) -> Result<(), ()> {
+        _ => return Ok(()),
+    };
+
+    if let Some(Expr::BinaryOp { right, .. }) = selection {
+        let right: &mut Expr = right;
+        replace_expr_with_params(right, params);
+    }
```
Could you please add handling of `left` parameter also?

alex-dukhno reviewed (changes requested):
> Is it a feature that change user experience?
> No

I think this feature change user experience but not by `psql` but rather with database driver.
Could you please add the following compatibility tests of executing:

`select * from schema_name.table_name where col1 > ?`
`select * from schema_name.table_name where ? > col1`
`select * from schema_name.table_name where ? > ?`

Thanks

silathdiir commented:
@alex-dukhno I tried to add the compatibility tests as you mentioned above. But it seems that the database driver sends Extended Protocol `Parse` with Parameter Type of OID zero (unspecified type) which we could not handle currently.
I tried to review the implementation of `materialize`. It inspects the real Parameter Type for OID zero in `select` plan.
I will convert this PR to draft and try to add this support in `query_planner`. Please inform me if you have any suggestions. Thanks.

silathdiir commented:
Hi @alex-dukhno , I fixed to support binding both left and right parameters of binary OP according to your review. Thanks.

And added cargo tests and compatibility tests. But I added the keyword `@ignore` for the three compatibility test cases. Since the Parameter Type of OID zero could not been handled by us now. Please help review if you have time. Thanks.

If you agree, I will create an issue for `query_analyzer` about analyzing this Parameter Type of OLD zero in Extended Protocol flow.

alex-dukhno commented:
> If you agree, I will create an issue for `query_analyzer` about analyzing this Parameter Type of OLD zero in Extended Protocol flow.

Previously we split any functionality by query types (selects, inserts, updates, deletes), lets try the other approach: split `query_analyzer` by possible places of parameters:
 * `insert into <table> values (?, ?, ?)` - assign operation
 * `update <table> set <column> = ?` - re-assign operation
 * `select * from <table>` - retrieve operation
 * `delete from <table> where <col> = ?`, `select * from <table> where <col> = ?`, `update <table> set <column> = x where <col2> = ?` - predicate operation

Let's create issue for each of them. If you think the list is not complete, feel free to suggest more items.

Thanks
